common/configuration/api.go

```diff
@@ -88,9 +90,38 @@ type Router[R Role] interface {
 type Provider[R Role] interface {
 	Role() R
 	Key() string
-	Load(ctx context.Context, ref Ref, key *url.URL) ([]byte, error)
+
 	// Store a configuration value and return its key.
 	Store(ctx context.Context, ref Ref, value []byte) (*url.URL, error)
 	// Delete a configuration value.
 	Delete(ctx context.Context, ref Ref) error
 }
+
+// SyncableProvider is an interface for providers that can load values on-demand.
+// This is recommended if the provider allows inexpensive loading of values.
+type OnDemandProvider[R Role] interface {
```

Comment:
These could be SynchronousProvider and AsynchronousProvider, but I'm just bike-shedding a bit.